File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/activity/award_type_controller.py
import inspect
import logging
from django.contrib import messages
from django.contrib.auth.decorators import login_required
from django.http import HttpResponse, HttpResponseRedirect, HttpRequest
from django.shortcuts import get_object_or_404, redirect, render
from django.utils.decorators import method_decorator
from django.views.decorators.http import require_http_methods
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from ....Models.activity.award_type import AwardType
from .._traits.controller import Controller
logger = logging.getLogger(__name__)
class AwardTypeController(Controller):

    @method_decorator(login_required)
    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        return (
            redirect(get_redirect_url(request))
            if not request.user.has_perm('manage award type')
            else render(
                request,
                'awardtype/index.html',
                {'awardtypes': AwardType.objects.filter(created_by=request.user.creator_id)},
            )
        )

    @method_decorator(login_required)
    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        return (
            redirect(get_redirect_url(request))
            if not request.user.has_perm('create award type')
            else render(request, 'awardtype/create.html')
        )

    @method_decorator(login_required)
    @classmethod
    @require_http_methods(['POST'])
    def store(cls, request) -> HttpResponseRedirect:
        CN = cls.__name__
        MN = inspect.currentframe().f_code.co_name

        if not request.user.has_perm('create award type'):
            return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)

        name = request.POST.get('name')
        if not name or len(name) > 20:
            messages.error(request, 'Name is required and must be at most 20 characters.')
            return redirect(get_redirect_url(request))

        try:
            AwardType.objects.create(name=name, created_by=request.user.creator_id)
            messages.success(request, 'AwardType successfully created.')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

        return redirect('awardtype.index')

    @method_decorator(login_required)
    @classmethod
    def show(cls, request, id: int) -> HttpResponseRedirect:
        return redirect('awardtype.index')  # TODO: consider handling not-found or permissions

    @method_decorator(login_required)
    @classmethod
    def edit(cls, request, id: int) -> HttpResponse:
        awardtype = get_object_or_404(AwardType, id=id)
        return (
            redirect(get_redirect_url(request))
            if not request.user.has_perm('edit award type') or awardtype.created_by != request.user.creator_id
            else render(request, 'awardtype/edit.html', {'awardtype': awardtype})
        )

    @method_decorator(login_required)
    @classmethod
    @require_http_methods(['POST'])
    def update(cls, request, id: int) -> HttpResponseRedirect:
        CN = cls.__name__
        MN = inspect.currentframe().f_code.co_name

        awardtype = get_object_or_404(AwardType, id=id)
        if not request.user.has_perm('edit award type') or awardtype.created_by != request.user.creator_id:
            return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)

        name = request.POST.get('name')
        if not name or len(name) > 20:
            messages.error(request, 'Name is required and must be at most 20 characters.')
            return redirect(get_redirect_url(request))

        try:
            awardtype.name = name
            awardtype.save()
            messages.success(request, 'AwardType successfully updated.')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

        return redirect('awardtype.index')

    @method_decorator(login_required)
    @classmethod
    def destroy(cls, request, id: int) -> HttpResponseRedirect:
        CN = cls.__name__
        MN = inspect.currentframe().f_code.co_name

        awardtype = get_object_or_404(AwardType, id=id)
        if not request.user.has_perm('delete award type') or awardtype.created_by != request.user.creator_id:
            return default_permission_denial(request, err=None, ref=f'{CN}::{MN}', logger=logger)

        try:
            awardtype.delete()
            messages.success(request, 'AwardType successfully deleted.')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

        return redirect('awardtype.index')

File: _inc/laravel/app/Http/Controllers/Activity/AwardTypeController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{DatabaseConstants, UsersConstants};
use App\Models\AwardType;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Log, Route, Validator, View as ViewFacade};
use Illuminate\Support\Str;

class AwardTypeController extends Controller
{

  public function index(Request $request): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    $viewPath = Str::snake(Str::replaceLast('Controller', '', class_basename($class))) . '.index';
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $req->user()->id]);
        self::_setAuth($req, 'manage award type');
        $loadStart = microtime(true);
        $awardTypes = AwardType::query()->where(DatabaseConstants::COL_TABLE_CREATOR, $req->user()->creatorId())->get();
        $this->logExecutionTime($loadStart, $action, 'loadAwardTypes');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, ['awardTypes' => $awardTypes]);
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function create(Request $request): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    $viewPath = Str::snake(Str::replaceLast('Controller', '', class_basename($class))) . '.create';
    return $this->measureProfile($action, function () use ($req, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", [UsersConstants::COL_USER_ID => $req->user()->id]);
        self::_setAuth($req, 'create award type');
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath);
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function store(Request $request): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      try {
        self::_setAuth($req, 'create award type');
        $valStart = microtime(true);
        $validator = Validator::make($req->all(), ['name' => 'required|max:20']);
        if ($validator->fails()) return redirect()->back()->with('error', $validator->errors()->first());
        $this->logExecutionTime($valStart, $action, 'validateStore');
        $awardType = new AwardType();
        $awardType->name = $req->input('name');
        $awardType->created_by = $req->user()->creatorId();
        $saveStart = microtime(true);
        $awardType->save();
        $this->logExecutionTime($saveStart, $action, 'saveAwardType');
        $resource = Str::snake(Str::replaceLast('Controller', '', class_basename($class)));
        return redirect()->route($resource . '.index')->with('success', __(':resource successfully created.', ['resource' => Str::of($resource)->replace('_', ' ')->title()]));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class]);
  }

  public function show(Request $request, AwardType $awardType): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $action, $class) {
      try {
        $resource = Str::snake(Str::replaceLast('Controller', '', class_basename($class)));
        return redirect()->route($resource . '.index');
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'award_type_id' => $awardType->id]);
  }

  public function edit(Request $request, AwardType $awardType): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    $viewPath = Str::snake(Str::replaceLast('Controller', '', class_basename($class))) . '.edit';
    return $this->measureProfile($action, function () use ($req, $awardType, $action, $class, $viewPath) {
      try {
        Log::info("[{$class}::{$action}] start", ['award_type_id' => $awardType->id, UsersConstants::COL_USER_ID => $req->user()->id]);
        self::_setAuth($req, 'edit award type');
        if ($awardType->created_by !== $req->user()->creatorId()) throw new AuthorizationException;
        if (!ViewFacade::exists($viewPath)) return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
        return view($viewPath, ['awardType' => $awardType]);
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'award_type_id' => $awardType->id]);
  }

  public function update(Request $request, AwardType $awardType): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $awardType, $action, $class) {
      try {
        self::_setAuth($req, 'edit award type');
        if ($awardType->created_by !== $req->user()->creatorId()) throw new AuthorizationException;
        $valStart = microtime(true);
        $validator = Validator::make($req->all(), ['name' => 'required|max:20']);
        if ($validator->fails()) return redirect()->back()->with('error', $validator->errors()->first());
        $this->logExecutionTime($valStart, $action, 'validateUpdate');
        $awardType->name = $req->input('name');
        $saveStart = microtime(true);
        $awardType->save();
        $this->logExecutionTime($saveStart, $action, 'saveAwardType');
        $resource = Str::snake(Str::replaceLast('Controller', '', class_basename($class)));
        return redirect()->route($resource . '.index')->with('success', __(':resource successfully updated.', ['resource' => Str::of($resource)->replace('_', ' ')->title()]));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'award_type_id' => $awardType->id]);
  }

  public function destroy(Request $request, AwardType $awardType): mixed
  {
    $action = __FUNCTION__;
    $method = __METHOD__;
    $class = static::class;
    $req = $request;
    return $this->measureProfile($action, function () use ($req, $awardType, $action, $class) {
      try {
        self::_setAuth($req, 'delete award type');
        if ($awardType->created_by !== $req->user()->creatorId()) throw new AuthorizationException;
        $delStart = microtime(true);
        $awardType->delete();
        $this->logExecutionTime($delStart, $action, 'deleteAwardType');
        $resource = Str::snake(Str::replaceLast('Controller', '', class_basename($class)));
        return redirect()->route($resource . '.index')->with('success', __(':resource successfully deleted.', ['resource' => Str::of($resource)->replace('_', ' ')->title()]));
      } catch (AuthorizationException $e) {
        return defaultPermissionDenial($req, $e, $class . '::' . $action);
      } catch (\Throwable $e) {
        return defaultUndefinedException($req, $e, $class . '::' . $action);
      }
    }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $class, 'award_type_id' => $awardType->id]);
  }

  protected static function _setAuth(Request $request, string $permission, string $suffix = ''): void
  {
    $perm = $permission . ($suffix ? " {$suffix}" : '');
    if (!$request->user()->can($perm)) {
      throw new AuthorizationException;
    }
  }
}
